### PR TITLE
D3ASIM-3028: Added retry support for the websocket connection for the…

### PR DESCRIPTION
**Pull request by spyrostz**
… api client. Added test coverage.

### CODE REVIEW

File: d3a_api_client/websocket_device.py [hunk 1]
```diff
@@ -55,6 +61,21 @@ async def websocket_coroutine(websocket_uri, websocket_headers, message_dispatch
...
 
+async def retry_coroutine(websocket_uri, websocket_headers, message_dispatcher, retry_count=0):
+    ws_connect_time = time()
+    try:
+        await websocket_coroutine(websocket_uri, websocket_headers, message_dispatcher)
+    except Exception as e:
+        logging.warning(f"Connection failed, trying to reconnect.")
+        ws_error_time = time()
+        if ws_error_time - ws_connect_time > WEBSOCKET_ERROR_THRESHOLD_SECONDS:
+            retry_count = 0
```
Comment (hannesdiedrich) [inline d3a_api_client/websocket_device.py:72]: Can you please explain what you try to achieve here? 
Also, i dont see the retry_timeout increased somewhere. In the unittest you are mocking it with a global variable `coro_execution_counter` so it indeed gets increased, but not here. Or don't i see the forrest behind all the trees ? ;-)

**Review by hannesdiedrich — COMMENTED** [review]
One comment. Will review after addressed.

File: d3a_api_client/websocket_device.py (hunk 1 above)
Comment (hannesdiedrich) [reply d3a_api_client/websocket_device.py:72]: Alright, i missed the last line actually. 
But I have another question: if the `connection manages to be established`, no Exception would be raised in `websocket_coroutine` and this part of the code would not be run, right?

Comment (spyrostz) [reply d3a_api_client/websocket_device.py:72]: No exception will be raised indeed if the connection never crashes (in which case we do not need the retry), therefore all will go well. If, after a couple of hours though, the connection crashes again, an exception will be raised and the retry counter needs to start from 0 and not from the last value of the retry_count. In order to differentiate if the connect has succeeded for some time or not, the easiest way is to use a timeout for that (WEBSOCKET_ERROR_THRESHOLD_SECONDS) in our case.

Comment (hannesdiedrich) [reply d3a_api_client/websocket_device.py:72]: AH, ok, makes sense, thanks for being patient with me!

**Review by hannesdiedrich — APPROVED** [review]
LGTM

File: d3a_api_client/websocket_device.py (hunk 1 above)
Comment (spyrostz) [reply d3a_api_client/websocket_device.py:72]: No worries, thanks!

**Review by faizan2590 — APPROVED** [review]
LGTM

**Pull request merged**